Remove hard-coded colors from output, introduce additional $reportthemes

asears opened the pull request:
<!--

Thank you for contributing to Pester!

-->

## PR Summary

Remove hard-coded color names from Output.ps1.  Introduce PR for suggestions of pluggable theme & splitting up failure color with failure message color.  #1735 relates.  

## PR Checklist

- [x] PR has meaningful title
- [x] Summary describes changes
- [ ] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [ ] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

fflaten commented on src/functions/Output.ps1:
```diff
@@ -58,6 +59,12 @@ $script:ReportTheme = DATA {
         Information      = 'DarkGray'
         Coverage         = 'White'
         CoverageWarn     = 'DarkRed'
+        DiscoveryStart   = 'Magenta'
+        DiscoveryFilter  = 'Magenta'
+        Discovery        = 'Magenta'
+        DiscoveryFinish  = 'Magenta'
```
Seems a bit overkill. Would `Discovery` be enough?

fflaten commented:
As "FailMessage" looks to be put on hold for now, maybe remove it so this can be merge for hardcode-cleanup?

Do we need `TestFail` when it's the same as `Fail`? I would just reuse Fail for now.
Not sure about `TestRunBlock`, but `Discovery` feels wrong too. Why is it Magenta in the first place, would've thought Foreground as it's neutral information like "tests completed .."?

@nohwnd is the original artist here. What do you think?

nohwnd commented on src/functions/Output.ps1:
```diff
@@ -698,7 +702,7 @@ function Get-WriteScreenPlugin ($Verbosity) {
 
         foreach ($e in $Context.Block.ErrorRecord) { ConvertTo-FailureLines $e }
-        & $SafeCommands['Write-Host'] -ForegroundColor Red "[-] $($Context.Block.FrameworkData.CommandUsed) $($Context.Block.Path -join ".") failed"
+        & $SafeCommands['Write-Host'] -ForegroundColor $ReportTheme.TestFail "[-] $($Context.Block.FrameworkData.CommandUsed) $($Context.Block.Path -join ".") failed"
```
This is a failed block not a failed test. This would happen when you have failure in AfterAll teardown for example.

asears replied:
renamed to BlockFail

nohwnd commented on src/functions/Output.ps1:
```diff
@@ -507,7 +511,7 @@ function Get-WriteScreenPlugin ($Verbosity) {
             if ("file" -eq $Context.Block.BlockContainer.Type) {
                 # write two spaces to separate each file
-                & $SafeCommands["Write-Host"] -ForegroundColor Magenta "`nRunning tests from '$($Context.Block.BlockContainer.Item)'"
+                & $SafeCommands["Write-Host"] -ForegroundColor $ReportTheme.TestRunBlock "`nRunning tests from '$($Context.Block.BlockContainer.Item)'"
```
This is a container run. Can we name it ContainerRun, or just Container?

asears replied:
renamed to container

nohwnd commented on src/functions/Output.ps1:
```diff
@@ -494,11 +498,11 @@ function Get-WriteScreenPlugin ($Verbosity) {
         # if ($Context.AnyFocusedTests) {
         #     $focusedTests = $Context.FocusedTests
-        #     & $SafeCommands["Write-Host"] -ForegroundColor Magenta "There are some ($($focusedTests.Count)) focused tests '$($(foreach ($p in $focusedTests) { $p -join "." }) -join ",")' running just them."
+        #     & $SafeCommands["Write-Host"] -ForegroundColor $ReportTheme.TestRunBlock "There are some ($($focusedTests.Count)) focused tests '$($(foreach ($p in $focusedTests) { $p -join "." }) -join ",")' running just them."
```
This is commented out and won't be used. Probably Find&Replace error?

asears replied:
removed

nohwnd commented on src/functions/Output.ps1:
```diff
@@ -43,6 +43,7 @@ $script:ReportTheme = DATA {
         PassTime         = 'DarkGray'
         Fail             = 'Red'
         FailTime         = 'DarkGray'
+        FailMessage      = 'Gray'
```
If the suggestions of what to color darker are implemented this would be stack trace. Maybe just call it FailDetail?

asears replied:
changed

nohwnd commented:
I don't remember why it is magenta, probably because I wanted to have distinct colors for each phase while I was developing it, or watched too many @danielbohannon talks and got inspired. Who knows.

I am open to changes as long as they bring improvements. 

There actuall were multiple color profiles in the past, when using PowerGUI for powershell was a thing, there you had white background and so I added light and dark theme which was then later removed when Gherkin was merged into Pester.

asears commented:
> I don't remember why it is magenta, probably because I wanted to have distinct colors for each phase while I was developing it, or watched too many @danielbohannon talks and got inspired. Who knows.
> 
> I am open to changes as long as they bring improvements.
> 
> There actuall were multiple color profiles in the past, when using PowerGUI for powershell was a thing, there you had white background and so I added light and dark theme which was then later removed when Gherkin was merged into Pester.

PowerGUI, that brought back some memories!  What a great tool from @KirkMunro and team...  Thanks for sharing Daniel's info and why you chose this.  Color schemes are hard, especially now that Powershell Core supports Linux and they vary through terminal windows, VSCode and in web browsers/Devops.  

I've incorporated the feedback and barring any git issues should be in the branch.  Will need some more conversation and another PR before we introduce the original failmessage change I suppose?
